modin/core/execution/dask/implementations/pandas_on_dask/partitioning/virtual_partition.py

```diff
@@ -46,65 +47,76 @@ class PandasOnDaskDataframeVirtualPartition(PandasDataframeAxisPartition):
-                ) else new_list_of_blocks.append(
-                    obj
-                )
-            self.list_of_partitions_to_combine = new_list_of_blocks
-        # Materialize partitions if the axis of this virtual does not match the virtual partitions
-        else:
-            self.list_of_partitions_to_combine = [
-                obj.force_materialization().list_of_partitions_to_combine[0]
-                if isinstance(obj, PandasOnDaskDataframeVirtualPartition)
-                else obj
-                for obj in list_of_blocks
-            ]
+        self._list_of_constituent_partitions = list_of_partitions
+        # Defer computing _list_of_block_partitions because we might need to
+        # drain call queues for that.
+        self._list_of_block_partitions = None
 
     partition_type = PandasOnDaskDataframePartition
     instance_type = Future
 
+    @property
+    def list_of_block_partitions(self) -> List[partition_type]:
+        """
+        Get the list of block partitions that compose this partition.
+
+        Returns
+        -------
+        List
+            A list of ``PandasOnDaskDataframePartition``.
+        """
+        if self._list_of_block_partitions is not None:
+            return self._list_of_block_partitions
+        self._list_of_block_partitions = []
+        # Extract block partitions from the block and virtual partitions that
+        # constitute this partition.
+        for partition in self._list_of_constituent_partitions:
+            if isinstance(partition, PandasOnDaskDataframeVirtualPartition):
+                if partition.axis == self.axis:
+                    # We are building a virtual partition out of another
+                    # virtual partition `partition` that contains its own list
+                    # of block partitions, partition.list_of_block_partitions.
+                    # `partition` may have its own call queue, which has to be
+                    # applied to the entire `partition` before we execute any
+                    # further operations on its block parittions.
+                    partition.drain_call_queue()
+                    self._list_of_block_partitions.extend(
+                        partition.list_of_block_partitions
+                    )
+                else:
+                    # If this virtual partition is made of virtual partitions
+                    # for the other axes, squeeze such partitions into a single
+                    # block so that this partition only holds a one-dimensional
+                    # list of blocks. We could change this implementation to
+                    # hold a 2-d list of blocks, but that would complicate the
+                    # code quite a bit.
+                    self._list_of_block_partitions.append(
+                        partition.force_materialization().list_of_block_partitions[0]
+                    )
+            else:
+                self._list_of_block_partitions.append(partition)
```

Comment:
I just checked and apply does not - that means there's a potential bug here if we have block partitions with a call queue. We could just resolve that by draining call queue in apply (so that way it's lazy). Up to you whether that belongs in this PR or another

Reply:
`apply` always [calls](https://github.com/modin-project/modin/blob/7a36071c0b00e0392615a0dd9d5c2ddd5f7c0d27/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/virtual_partition.py#L301) superclass apply, which only ever accesses [`list_of_blocks`](https://github.com/modin-project/modin/blob/b22b93df20ad25ae7a11f0c89d32fb2f234d4641/modin/core/dataframe/pandas/partitioning/axis_partition.py#L66-L99). I also tested for this case specifically in `test_from_block_and_virtual_partition_with_call_queues `. Anyway, if anyone (including this class) calls `apply` on an undrained block from a virtual partition (as `modin/experimental/batch/pipeline.py` can), the block will drain its own call queue.